metrics/README.md

```diff
@@ -0,0 +1,99 @@
+# Common metrics export interfaces for Knative
+
+_Note that this directory is currently in transition. See [the Plan](#the-plan)
+for details on where this is heading._
+
+## Current status
+
+The code currently uses OpenCensus to support exporting metrics to multiple
+backends. Currently, two backends are supported: Prometheus and Stackdriver.
+
+Metrics export is controlled by a ConfigMap called `config-observability` which
+is a key-value map with specific values supported for each of the Stackdriver
+and Prometheus backends. Hot-reload of the ConfigMap on a running process is
+supported by directly watching (via the Kubernetes API) the
+`config-observability` object. Configuration via environment is also supported
+for use by the `queue-proxy`, which runs with user permissions in the user's
+namespace.
+
+## Problems
+
+There are currently
+[6 supported Golang exporters for OpenCensus](https://opencensus.io/exporters/supported-exporters/go/).
+At least the Stackdriver exporter causes problems/failures if started without
+access to (Google) application default credentials. It's not clear that we want
+to build all of those backends into the core of `knative.dev/pkg` and all
+downstream dependents, and we'd like all the code shipped in `knative.dev/pkg`
+to be able to be tested without needing any special environment setup.
+
+With the current direct-integration setup, there needs to be initial and ongoing
+work in `pkg` (which should be high-value, low-churn code) to maintain and
+update stats exporters which need to be statically linked into ~all Knative
+binaries. This setup also causes problems for vendors who may want or need to
+perform an out-of-tree integration (e.g. proprietary or partially-proprietary
+monitoring stacks).
+
+Another problem is that each vendor's exporter requires different parameters,
+supplied as Golang `Options` methods which may require complex connections with
+the Knative ConfigMap. Two examples of this are secrets like API keys and the
+Prometheus monitoring port (which requires additional service/etc wiring).
+
+See also
+[this doc](https://docs.google.com/document/d/1t-aov3XrhobjCKW4kwScY44QAoahiwxoyXXFtZyL8jw/edit),
+where the plan was worked out.
+
+## The plan
+
+OpenCensus (and eventually OpenTelemetry) offers an sidecar or host-level agent
+with speaks the OpenCensus protocol and can proxy from this protocol to multiple
+backends.
+
+![OpenCensus Agent configuration](https://github.com/census-instrumentation/opencensus-service/raw/master/images/opencensus-service-deployment-models.png)
+(From OpenCensus Documentation)
+
+**We will standardize on export to the OpenCensus export protocol, and encourage
```

Comment:
I think this is more accurately standardizing on the OpenCensus client library? Exporting to be handled by some other OpenCensus component, either agent or collector service (though IMO, we should not support agent).

Reply:
No, the standardization is on the gRPC and/or REST/JSON API exposed by the the OpenCensus Agent and Collector. The API is documented here: https://github.com/census-instrumentation/opencensus-service/blob/master/DESIGN.md#agent-communication and here: https://github.com/census-instrumentation/opencensus-proto/tree/master/src/opencensus/proto/agent

Reply:
Nvm, found some code samples, OpenCensus exporter is the way to go.

Reply:
Ah I was late to see your response too :)